### PR TITLE
Fix camera jumps with terrain and transformCameraUpdate.

### LINKED ISSUE #4233: transformCameraUpdate leads to unexpected camera jumps when targetting terrain
**maplibre-gl-js version**: 4.3.2

**browser**: Firefox 126, Chrome 125

### Steps to Trigger Behavior

 1. Use the 3D-terrain example
 2. Add a transformCameraUpdate function in the options that returns the input or an empty object.
 3. Drag pan the map a little bit (inertia must be triggered). After end of drag, the jump happens.
 4. After the jump, it works fine until the terrain is targetted again with the camera.

### Link to Demonstration

https://jsbin.com/pobidixeqa/1/edit?html,output

### Expected Behavior

The empty `((tr)=>({})` or identity `(tr)=>tr` transformCameraUpdate option should never have an effect on the camera movement.

### Actual Behavior

The camera jumps to a different location when drag panning with triggered inertia and terrain targeted.

### PR DESCRIPTION
**Pull request by chrneumann**
Fixes camera jumps at end of movement when terrain and transformCameraUpdate is enabled (fixes #4233).

Had to enable WebGL for headless Chrome, as I could not reproduce the bug without it in headless mode.
The problem was caused by directly accessing the map's transform, instead of using _getTransformForUpdate and _applyUpdatedTransform.

## Launch Checklist

 - [x] Confirm **your changes do not include backports from Mapbox projects** (unless with compliant license) - if you are not sure about this, please ask!
 - [X] Briefly describe the changes in this PR.
 - [x] Link to related issues.
 - [X] Include before/after visuals or gifs if this PR includes visual changes.
 - [X] Write tests for all new functionality.
 - [X] Document any changes to public APIs.
 - [ ] Post benchmark scores.
 - [ ] Add an entry to `CHANGELOG.md` under the `## main` section.

### CODE REVIEW

File: test/integration/browser/browser.test.ts [hunk 1]
```diff
@@ -400,4 +406,76 @@ describe('Browser tests', () => {
...
+                        }
+                    ],
+                    terrain: {
+                        source: 'terrainSource',
+                        exaggeration: 1
+                    }
+                });
+
+            return new Promise<any>((resolve) => {
+                map.once('idle', () => {
```
Comment (HarelM) [inline test/integration/browser/browser.test.ts:458]: You can use `await map.once('idle');`.

File: test/integration/browser/browser.test.ts [hunk 2]
```diff
@@ -400,4 +406,76 @@ describe('Browser tests', () => {
         await expect(rtlPromise).rejects.toThrow(regex);
 
     }, 2000);
+
+    test('Movement with transformCameraUpdate and terrain', async () => {
+        await page.evaluate(() => {
+            map.setPitch(52)
+                .setZoom(15)
+                .setCenter([11.40, 47.30])
+                .setStyle({
```
Comment (HarelM) [inline test/integration/browser/browser.test.ts:415]: Can we use a simpler style for the test?
Can we use a different type of test maybe?
This is more of a e2e test, which takes long to run, can we use a unit-test instead?

File: test/integration/browser/browser.test.ts [hunk 3]
```diff
@@ -26,7 +26,13 @@ describe('Browser tests', () => {
         );
         await new Promise<void>((resolve) => server.listen(resolve));
 
-        browser = await puppeteer.launch({headless: true});
+        browser = await puppeteer.launch({
+            headless: true,
+            args: [
```
Comment (HarelM) [inline test/integration/browser/browser.test.ts:31]: Do the test fail if these parameters aren't used?

**Comment by HarelM** [comment]
Thanks for taking the time to fix this issue!
This looks like the right solution to the problem.
See my small comments above.

File: test/integration/browser/browser.test.ts (hunk 3 above)
Comment (chrneumann) [reply test/integration/browser/browser.test.ts:31]: The test will *never* fail if not used.

File: test/integration/browser/browser.test.ts (hunk 2 above)
Comment (chrneumann) [reply test/integration/browser/browser.test.ts:415]: Tried it, but could not reproduce it with a unit test.

Comment (chrneumann) [reply test/integration/browser/browser.test.ts:415]: No, not with simply adding the transformCameraUpdate option. As said, the browser test does not even fail with WebGL disabled. I don't know why, I can't say that I understand every bit of the changes involved (as said elsewhere, it's hard to figure out without use case examples for the transformCameraUpdate stuff).

Comment (UberMouse) [reply test/integration/browser/browser.test.ts:415]: Here is our `transformCameraUpdate` usage as an example

```
map.transformCameraUpdate = (next) => {
	return {
    	...next,
        bearing: displaySettings.rotationLock ?? next.bearing,
        pitch: displaySettings.tiltLock ?? next.pitch,
    };
};
```

This is to enable us to lock the camera to a specific bearing and/or pitch based on settings. Turning off interactions isn't quite foolproof enough, but this locks it perfectly.

**Comment by HarelM** [comment]
I think the following unit test can reproduce this issue to some extent - as the recalculate zoom will be done on the original transform and won't take the zoom changes that are done as part of the transfrom camera update:
```ts
    test('recalculate zoom is done on the camera update transform', () => {
        const map = createMap({
            interactive: true, 
            clickTolerance: 4,
            transformCameraUpdate: ({zoom}) => ({ zoom: zoom + 0.1})
        });
        map.terrain = {
            pointCoordinate: () => null,
            getElevationForLngLatZoom: () => 1000
        } as any;
        const canvas = map.getCanvas();
        simulate.dragWithMove(canvas, {x: 100, y: 100}, {x: 100, y: 150});
        map._renderTaskQueue.run();

        expect(map.getZoom()).toBe(0.20008255564976615);
    });
```
What do you think?

**Comment by chrneumann** [comment]
Seems fine, still I think an integration test that involves transformCameraUpdate wouldn't be bad. But fine for me :)

**Comment by HarelM** [comment]
I'm good with both tests, just please simplify the style in the integration test...

**Comment by chrneumann** [comment]
> I'm good with both tests, just please simplify the style in the integration test...

Ah, I did that on purpose. Thought about simplifying it, which would be fine for the automatic testing, but if you want to observe the bug, the more complete style is really helpful. But yeah, I can simplify it if you don't think thats important.

**Comment by HarelM** [comment]
Yeah, the test should be as simple as possible, for debugging you can change the examples in order to reproduce an issue.
Please add the test I sent and simplify the integration test and I'll merge this PR.
Thanks!!

**Comment by chrneumann** [comment]
> I think the following unit test can reproduce this issue to some extent - as the recalculate zoom will be done on the original transform and won't take the zoom changes that are done as part of the transfrom camera update:
> What do you think?

Ah, you did not actually test it? At least, it never fails for me.

**Comment by HarelM** [comment]
I've test it with and without the fix and it gave different results.

**Comment by chrneumann** [comment]
Maybe I'm missing something. I pushed your unit test, you can check it with the commit "04278f129 | Fix terrain jumps with transformCameraUpdate (#4233)" reversed, which lets the integration test fail for me but not the unit test.

**Comment by chrneumann** [comment]
Ah okay, it fails when the fix is applied.

**Comment by chrneumann** [comment]
Ah, it fails when the fix is applied? Did you calculate the zoom by hand? Or is it fine to just fix the expected zoom value with the fix applied?

**Comment by HarelM** [comment]
Must have written the test when I removed the fix, sorry...

**Review by HarelM — APPROVED** [review]

**Pull request merged**